File: scripts/tags.lua
-- Copyright (C) 2018  David Capello
--
-- This file is released under the terms of the MIT license.
-- Read LICENSE.txt for more information.

do
  local s = Sprite(32, 32)
  for i = 1,7 do s:newFrame() end
  assert(#s.frames == 8)

  a = s:newTag(1, 2)
  b = s:newTag(3, 4)
  c = s:newTag(5, 8)

  assert(a.frames == 2)
  assert(b.frames == 2)
  assert(c.frames == 4)

  assert(a == s.tags[1])
  assert(b == s.tags[2])
  assert(c == s.tags[3])

  local i = 1
  for k,v in ipairs(s.tags) do
    assert(i == k)
    assert(v == s.tags[k])
    i = i+1
  end

  s:deleteTag(b)
  assert(a == s.tags[1])
  assert(c == s.tags[2])

  assert(c.fromFrame.frameNumber == 5)
  assert(c.toFrame.frameNumber == 8)
end

File: src/doc/tags.cpp
// Aseprite Document Library
// Copyright (C) 2019  Igara Studio S.A.
// Copyright (C) 2001-2016 David Capello
//
// This file is released under the terms of the MIT license.
// Read LICENSE.txt for more information.

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "doc/tags.h"

#include "base/debug.h"
#include "doc/tag.h"

#include <algorithm>

namespace doc {

Tags::Tags(Sprite* sprite)
  : m_sprite(sprite)
{
}

Tags::~Tags()
{
  for (Tag* tag : m_tags) {
    tag->setOwner(nullptr);
    delete tag;
  }
}

void Tags::add(Tag* tag)
{
  auto it = begin(), end = this->end();
  for (; it != end; ++it) {
    if ((*it)->fromFrame() > tag->fromFrame() ||
        ((*it)->fromFrame() == tag->fromFrame() &&
         (*it)->toFrame() < tag->toFrame()))
      break;
  }
  m_tags.insert(it, tag);
  tag->setOwner(this);
}

void Tags::remove(Tag* tag)
{
  auto it = std::find(m_tags.begin(), m_tags.end(), tag);
  ASSERT(it != m_tags.end());
  if (it != m_tags.end())
    m_tags.erase(it);

  tag->setOwner(nullptr);
}

Tag* Tags::getByName(const std::string& name) const
{
  for (Tag* tag : *this) {
    if (tag->name() == name)
      return tag;
  }
  return nullptr;
}

Tag* Tags::getById(ObjectId id) const
{
  for (Tag* tag : *this) {
    if (tag->id() == id)
      return tag;
  }
  return nullptr;
}

Tag* Tags::innerTag(const frame_t frame) const
{
  const Tag* found = nullptr;
  for (const Tag* tag : *this) {
    if (frame >= tag->fromFrame() &&
        frame <= tag->toFrame()) {
      if (!found ||
          (tag->toFrame() - tag->fromFrame()) < (found->toFrame() - found->fromFrame())) {
        found = tag;
      }
    }
  }
  return const_cast<Tag*>(found);
}

Tag* Tags::outerTag(const frame_t frame) const
{
  const Tag* found = nullptr;
  for (const Tag* tag : *this) {
    if (frame >= tag->fromFrame() &&
        frame <= tag->toFrame()) {
      if (!found ||
          (tag->toFrame() - tag->fromFrame()) > (found->toFrame() - found->fromFrame())) {
        found = tag;
      }
    }
  }
  return const_cast<Tag*>(found);
}

} // namespace doc
